web/src/lib/pagination.ts

```diff
@@ -1,3 +1,84 @@
+const DEFAULT_PAGE_SIZE = 10;
+
+export type Pagination = {
+	currentResultsPage: number;
+	resultsPerPage: number;
+	resultsToFetchStart: number;
+	resultsToFetchEnd: number;
+	resultShowingFirst: number;
+	resultShowingLast: number;
+	resultsCount: number;
+	pagesDisplay: Array<string>
+	pagesCount: number;
+	prevPageValid: boolean;
+	prevPageUrl: string | null;
+	nextPageValid: boolean;
+	nextPageUrl: string | null;
+};
+
+export const getPagePagination = (url: URL, resultsCount: number, resultsPerPage = DEFAULT_PAGE_SIZE ): Pagination => {
+
+	// from the given URL, grab which page of results is currently selected (default 1)
+	const currentResultsPage = parseInt(url.searchParams.get('page')) || 1;
+
+	// define the start / stop range of rows we want to retreive from the DB for the page
+	const resultsToFetchStart = (currentResultsPage - 1) * resultsPerPage;
+	const resultsToFetchEnd = resultsToFetchStart + resultsPerPage;
+
+	// of the results, the first / last displayed on the page
+	const resultShowingFirst = (currentResultsPage - 1) * resultsPerPage + 1;
+	const resultShowingLast = Math.min(
+		currentResultsPage * resultsPerPage,
+		resultsCount
+	);
+
+	// the total number of pages
+	const pagesCount = Math.ceil(resultsCount / resultsPerPage)
+
+	// determine, based on the current page, if the prev / next pages are valid
+	const prevPageValid = (currentResultsPage > 1);
+	const nextPageValid = (currentResultsPage < pagesCount);
+
+	// if the prev / next page are valid, store what page they should correspond to
+	let prevPageUrl = prevPageValid ? new URL(url.href) : null;
+	if (prevPageValid) {
+		prevPageUrl.searchParams.set('page', currentResultsPage - 1);
+		prevPageUrl = prevPageUrl.href
+	}
+
+	let nextPageUrl = nextPageValid ? new URL(url.href) : null;
+	if (nextPageValid) {
+		nextPageUrl.searchParams.set('page', currentResultsPage + 1);
+		nextPageUrl = nextPageUrl.href
+	}
```

Comment:
I think it's a bit more readable to not check if the page is valid twice
```suggestion
	// if the prev / next page are valid, store what page they should correspond to
	let prevPageUrl =  null;
	if (prevPageValid) {
	        prevPageUrl = new URL(url.href)
		prevPageUrl.searchParams.set('page', currentResultsPage - 1);
		prevPageUrl = prevPageUrl.href
	}

	let nextPageUrl = null;
	if (nextPageValid) {
	        nextPageUrl = new URL(url.href);
		nextPageUrl.searchParams.set('page', currentResultsPage + 1);
		nextPageUrl = nextPageUrl.href
	}
```

could also add a helper function to clean it up